File: ios/Helpers/CallsUtils.swift
//
//  CallsUtils.swift
//  RNSendbirdCalls
//
//  Created by Airen Kang on 2022/05/03.
//  Copyright © 2022 Sendbird. All rights reserved.
//

import Foundation
import SendBirdCalls
import React
import AVFoundation

// MARK: Converter
class CallsUtils {
    static func convertUserToDict(_ user: SendBirdCalls.User) -> [String: Any?] {
        return [
            "userId": user.userId,
            "isActive": user.isActive,
            "nickname": user.nickname ?? "",
            "metaData": user.metadata ?? [:],
            "profileUrl": user.profileURL ?? ""
        ]
    }
    
    static func convertDirectCallToDict(_ call: DirectCall) -> [String: Any?] {
        return [
            "callId": call.callId,
            "callUUID": call.callUUID?.uuidString,
            "callLog": convertDirectCallLogToDict(call.callLog),
            "callee": convertDirectCallUserToDict(call.callee),
            "caller": convertDirectCallUserToDict(call.caller),
            "endedBy": convertDirectCallUserToDict(call.endedBy),
            "customItems": call.customItems,
            "duration": call.duration,
            "endResult": call.endResult.asString,
            "localUser": convertDirectCallUserToDict(call.localUser),
            "remoteUser": convertDirectCallUserToDict(call.remoteUser),
            "myRole": call.myRole.asString,
            "availableVideoDevices": call.availableVideoDevices.map { convertVideoDeviceToDict(device: $0) },
            "currentVideoDevice": convertVideoDeviceToDict(device: call.currentVideoDevice),
            "isEnded": call.isEnded,
            "isOnHold": call.isOnHold,
            "isOngoing": call.isOngoing,
            "isVideoCall": call.isVideoCall,
            "isLocalScreenShareEnabled": call.isLocalScreenShareEnabled,
            "isLocalAudioEnabled": call.isLocalAudioEnabled,
            "isLocalVideoEnabled": call.isLocalVideoEnabled,
            "isRemoteAudioEnabled": call.isRemoteAudioEnabled,
            "isRemoteVideoEnabled": call.isRemoteVideoEnabled,
            "localRecordingStatus": call.localRecordingStatus.asString,
            "remoteRecordingStatus": call.remoteRecordingStatus.asString,
            
        ]
    }
    
    static func convertDirectCallLogToDict(_ callLog: DirectCallLog?) -> [String: Any?]? {
        guard let callLog = callLog else {
            return nil
        }
        
        return [
            "startedAt": callLog.startedAt,
            "endedAt": callLog.endedAt,
            "duration": callLog.duration,
            "callId": callLog.callId,
            "isFromServer": callLog.isFromServer,
            "isVideoCall": callLog.isVideoCall,
            "customItems": callLog.customItems,
            "endResult": callLog.endResult.asString,
            "myRole": callLog.myRole.asString,
            "callee": convertDirectCallUserToDict(callLog.callee),
            "caller": convertDirectCallUserToDict(callLog.caller),
            "endedBy": convertDirectCallUserToDict(callLog.endedBy),
        ]
    }
    
    static func convertDirectCallUserToDict(_ callUser: DirectCallUser?) -> [String: Any?]? {
        guard let callUser = callUser else {
            return nil
        }
        
        var user = convertUserToDict(callUser)
        user["role"] = callUser.role.asString
        return user
    }
    
    static func convertVideoDeviceToDict(device: VideoDevice?) -> [String: String]? {
        guard let device = device else {
            return nil
        }
        
        return [
            "deviceId": device.uniqueId,
            "position": device.position.asString
        ]
    }
    
    static func convertDictToCallOptions(_ options: [String: Any?]) -> CallOptions {
        let callOptions = CallOptions()
        
        if let audioEnabled = options["audioEnabled"] as? Bool {
            callOptions.isAudioEnabled = audioEnabled
        }
        if let videoEnabled = options["videoEnabled"] as? Bool {
            callOptions.isVideoEnabled = videoEnabled
        }
        if let frontCamera = options["frontCamera"] as? Bool {
            callOptions.useFrontCamera = frontCamera
        }
        if let localVideoViewId = options["localVideoViewId"] as? NSNumber {
            callOptions.localVideoView = safeGet {
                try CallsUtils.findViewBy(CallsEvents.shared.bridge, localVideoViewId).surface
            }
        }
        if let remoteVideoViewId = options["remoteVideoViewId"] as? NSNumber {
            callOptions.remoteVideoView = safeGet {
                try CallsUtils.findViewBy(CallsEvents.shared.bridge, remoteVideoViewId).surface
            }
        }
        
        return callOptions
    }
    
    static func convertAVRouteToDict(_ route: AVAudioSessionRouteDescription) -> [String: [[String: String]]] {
        return [
            "inputs": route.inputs.map {
                [
                    "name": $0.portName,
                    "type": $0.portType.asString()
                ]
            },
            "outputs": route.outputs.map {
                [
                    "name": $0.portName,
                    "type": $0.portType.asString()
                ]
            }
        ]
    }
    
    static func convertRoomToDict(_ room: Room?) -> [String: Any]? {
        guard let room = room else {
            return nil
        }
        
        return [
            "roomId": room.roomId,
            "state": room.state.asString,
            "type": room.type.asString,
            "customItems": room.customItems,
            
            "participants": room.participants.map { convertParticipantToDict($0) },
            "localParticipant": convertParticipantToDict(room.localParticipant)!,
            "remoteParticipants": room.remoteParticipants.map { convertParticipantToDict($0) },
            
            "createdAt": room.createdAt,
            "createdBy": room.createdBy
        ]
    }
    
    static func convertParticipantToDict(_ participant: Participant?) -> [String: Any]? {
        guard let participant = participant else {
            return nil
        }
        
        return [
            "participantId": participant.participantId,
            "user": convertUserToDict(participant.user),
            "state": participant.state.asString,
            
            "enteredAt": participant.enteredAt,
            "exitedAt": participant.exitedAt,
            "duration": participant.duration,
            
            "isAudioEnabled": participant.isAudioEnabled,
            "isVideoEnabled": participant.isVideoEnabled,
            
            "updatedAt": participant.updatedAt,
            // "videoView": participant.videoView
        ]
    }
}

// MARK: Safe runner block
extension CallsUtils {
    static func safeRun(completion: () throws -> Void) {
        do {
            try completion()
        } catch { }
    }
    
    static func safeGet<T>(_ completion: () throws -> T) -> T? {
        do {
            return try completion()
        } catch {
            return nil
        }
    }
}

// MARK: Finder
extension CallsUtils {
    static func findDirectCallBy(_ callIdOrUUID: String) throws -> DirectCall {
        if let callFromId = SendBirdCall.getCall(forCallId: callIdOrUUID) {
            return callFromId
        } else if let callFromUUID = SendBirdCall.getCall(forUUID: UUID(uuidString: callIdOrUUID)!) {
            return callFromUUID
        }
        throw RNCallsInternalError.notFoundDirectCall("findDirectCallBy")
    }
    
    static func findViewBy(_ bridge: RCTBridge, _ tag: NSNumber) throws -> BaseVideoView {
        if let view = bridge.uiManager.view(forReactTag: tag) as? BaseVideoView {
            return view
        }
        throw RNCallsInternalError.notFoundVideoView("findViewBy")
    }
}

File: android/src/main/java/com/sendbird/calls/reactnative/utils/CallsUtils.kt
package com.sendbird.calls.reactnative.utils

import android.app.ActivityManager
import android.app.ActivityManager.RunningAppProcessInfo
import android.content.Context
import android.util.Log
import com.facebook.react.bridge.*
import com.facebook.react.common.LifecycleState
import com.sendbird.calls.*
import com.sendbird.calls.reactnative.RNCallsInternalError
import com.sendbird.calls.reactnative.extension.asString
import com.sendbird.calls.reactnative.extension.rejectCalls
import com.sendbird.calls.reactnative.module.CallsModule
import com.sendbird.calls.reactnative.view.BaseVideoView

object CallsUtils {
    fun safeRun(fn: () -> Unit) {
        try {
            fn()
        } catch (e: Throwable) {
            Log.e(CallsModule.NAME, "[CallsUtils.safeRun] Catch error -> $e")
        }
    }
    fun safeRun(promise: Promise, fn: () -> Unit) {
        try {
            fn()
        } catch (e: Throwable) {
            when(e){
                is SendBirdException -> promise.rejectCalls(e)
                is RNCallsInternalError -> promise.rejectCalls(e)
                else -> promise.reject(e)
            }
        }
    }
    fun <T> safeGet(fn: () -> T): T? {
        return try {
            fn()
        } catch (e: Throwable){
            null
        }
    }

    fun findDirectCall(callId: String, from: String?): DirectCall {
        return SendBirdCall.getCall(callId) ?: throw RNCallsInternalError(from, RNCallsInternalError.Type.NOT_FOUND_DIRECT_CALL)
    }
    fun findVideoView(context: ReactContext, viewId: Int, from: String?): BaseVideoView {
        return context.currentActivity?.findViewById(viewId) ?: throw RNCallsInternalError(from, RNCallsInternalError.Type.NOT_FOUND_VIDEO_VIEW)
    }

    /** Data conversion utils **/
    fun insertMap(jsMap: WritableMap, key: String, value: Any?) = when (value) {
        is Boolean -> jsMap.putBoolean(key, value)
        is String -> jsMap.putString(key, value)
        is Double -> jsMap.putDouble(key, value)
        is Float -> jsMap.putDouble(key, value.toDouble())
        is Long -> jsMap.putDouble(key, value.toDouble())
        is Int -> jsMap.putInt(key, value)

        is Map<*, *> -> jsMap.putMap(key, convertToJsMap(value))
        is Array<*> -> jsMap.putArray(key, convertToJsArray(value))
        is List<*> -> jsMap.putArray(key, convertToJsArray(value))

        is WritableNativeMap -> jsMap.putMap(key, value)
        is ReadableNativeMap -> jsMap.putMap(key, value)
        is WritableNativeArray -> jsMap.putArray(key, value)
        is ReadableNativeArray -> jsMap.putArray(key, value)
        else -> {
            if(value == null) jsMap.putNull(key)
            else throw Error("Unknown type")
        }
    }
    fun insertArr(jsArr: WritableArray, value: Any?) = when (value) {
        is Boolean -> jsArr.pushBoolean(value)
        is String -> jsArr.pushString(value)
        is Double -> jsArr.pushDouble(value)
        is Float -> jsArr.pushDouble(value.toDouble())
        is Long -> jsArr.pushDouble(value.toDouble())
        is Int -> jsArr.pushInt(value)

        is Map<*, *> -> jsArr.pushMap(convertToJsMap(value))
        is Array<*> -> jsArr.pushArray(convertToJsArray(value))
        is List<*> -> jsArr.pushArray(convertToJsArray(value))

        is WritableNativeMap -> jsArr.pushMap(value)
        is ReadableNativeMap -> jsArr.pushMap(value)
        is WritableNativeArray -> jsArr.pushArray(value)
        is ReadableNativeArray -> jsArr.pushArray(value)
        else -> {
            if(value == null) jsArr.pushNull()
            else throw Error("Unknown type")
        }
    }
    fun convertToJsMap(nativeMap: Map<*, *>): WritableNativeMap {
        val jsMap = WritableNativeMap()
        val keyIterator = nativeMap.keys.iterator();
        while(keyIterator.hasNext()) {
            val key = keyIterator.next()
            val value = nativeMap.get(key)
            if(key is String) insertMap(jsMap, key, value)
        }
        return jsMap
    }
    fun convertToJsArray(nativeList: List<*>): WritableNativeArray {
        val jsArr = WritableNativeArray()
        val iterator = nativeList.iterator()
        while (iterator.hasNext()) {
            val value = iterator.next()
            insertArr(jsArr, value)
        }
        return jsArr
    }
    fun convertToJsArray(nativeArr: Array<*>): WritableNativeArray {
        val jsArr = WritableNativeArray()
        val iterator = nativeArr.iterator()
        while (iterator.hasNext()) {
            val value = iterator.next()
            insertArr(jsArr, value)
        }
        return jsArr
    }
    fun convertUserToJsMap(user: User) = convertToJsMap(mapOf(
        "userId" to user.userId,
        "isActive" to user.isActive,
        "nickname" to (user.nickname ?: ""),
        "metaData" to (user.metaData ?: WritableNativeMap()),
        "profileUrl" to (user.profileUrl ?: "")
    ))
    fun convertDirectCallToJsMap(call: DirectCall) = convertToJsMap(mapOf(
        "callId" to call.callId,
        "callLog" to when(call.callLog) {
            null -> null
            else -> convertDirectCallLogToJsMap(call.callLog!!)
        },
        "callee" to convertDirectCallUserToJsMap(call.callee),
        "caller" to convertDirectCallUserToJsMap(call.caller),
        "endedBy" to convertDirectCallUserToJsMap(call.endedBy),
        "customItems" to call.customItems,
        "duration" to call.duration,
        "endResult" to call.endResult.asString(),
        "localUser" to convertDirectCallUserToJsMap(call.localUser),
        "remoteUser" to convertDirectCallUserToJsMap(call.remoteUser),
        "myRole" to call.myRole!!.asString(),
        "availableVideoDevices" to call.availableVideoDevices.map { convertVideoDeviceToJsMap(it) },
        "currentVideoDevice" to convertVideoDeviceToJsMap(call.currentVideoDevice),
        "availableAudioDevices" to call.availableAudioDevices.map { it.asString() },
        "currentAudioDevice" to call.currentAudioDevice?.asString(),
        "isEnded" to call.isEnded,
        "isOnHold" to call.isOnHold,
        "isOngoing" to call.isOngoing,
        "isVideoCall" to call.isVideoCall,
        "isLocalScreenShareEnabled" to call.isLocalScreenShareEnabled,
        "isLocalAudioEnabled" to call.isLocalAudioEnabled,
        "isLocalVideoEnabled" to call.isLocalVideoEnabled,
        "isRemoteAudioEnabled" to call.isRemoteAudioEnabled,
        "isRemoteVideoEnabled" to call.isRemoteVideoEnabled,
        "localRecordingStatus" to call.localRecordingStatus.asString(),
        "remoteRecordingStatus" to call.remoteRecordingStatus.asString(),
    ))

    fun convertDirectCallLogToJsMap(callLog: DirectCallLog) = convertToJsMap(mapOf(
        "startedAt" to callLog.startedAt,
        "endedAt" to callLog.endedAt,
        "duration" to callLog.duration,
        "callId" to callLog.callId,
        "isFromServer" to callLog.isFromServer,
        "isVideoCall" to callLog.isVideoCall,
        "customItems" to callLog.customItems,
        "endResult" to callLog.endResult.asString(),
        "myRole" to callLog.myRole!!.asString(),
        "callee" to convertDirectCallUserToJsMap(callLog.callee),
        "caller" to convertDirectCallUserToJsMap(callLog.caller),
        "endedBy" to convertDirectCallUserToJsMap(callLog.endedBy),
    ))

    fun convertDirectCallUserToJsMap(callUser: DirectCallUser?) = when (callUser) {
        null -> null
        else -> {
            val user = convertUserToJsMap(callUser)
            insertMap(user, "role", callUser.role.asString())
            user
        }
    }

    fun convertVideoDeviceToJsMap(device: VideoDevice?) = when(device) {
        null -> null
        else -> convertToJsMap(mapOf(
            "deviceId" to device.deviceName,
            "position" to device.position.asString()
        ))
    }

    fun isAppInForeground(context: Context): Boolean {
        val activityManager =
            context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        val appProcesses = activityManager.runningAppProcesses ?: return false

        // Check if current activity is a background activity
        val packageName = context.packageName
        for (appProcess in appProcesses) {
            if (appProcess.importance == RunningAppProcessInfo.IMPORTANCE_FOREGROUND
                && appProcess.processName == packageName
            ) {
                val reactContext: ReactContext = try {
                    context as ReactContext
                } catch (exception: ClassCastException) {
                    // Not react context so default to true
                    return true
                }
                return reactContext.lifecycleState == LifecycleState.RESUMED
            }
        }
        return false
    }

    fun convertParticipantToJsMap(participant: Participant?) = when(participant) {
        null -> null
        else -> convertToJsMap(mapOf(
            "participantId" to participant.participantId,
            "user" to convertUserToJsMap(participant.user),
            "state" to participant.state.asString(),

            "enteredAt" to participant.enteredAt,
            "exitedAt" to (participant.exitedAt ?: 0),
            "duration" to (participant.duration ?: 0),

            "isAudioEnabled" to participant.isAudioEnabled,
            "isVideoEnabled" to participant.isVideoEnabled,

            "updatedAt" to participant.updatedAt,
        ))
    }

    fun convertRoomToJsMap(room: Room) = convertToJsMap(mapOf(
        "roomId" to room.roomId,
        "state" to room.state.asString(),
        "type" to room.type.asString(),
        "customItems" to room.customItems,

        "participants" to room.participants.map{ convertParticipantToJsMap(it) },
        "localParticipant" to convertParticipantToJsMap(room.localParticipant),
        "remoteParticipants" to room.remoteParticipants.map { convertParticipantToJsMap(it) },

        "availableAudioDevices" to room.availableAudioDevices.map { it.asString() },
        "currentAudioDevice" to room.currentAudioDevice?.asString(),

        "createdAt" to room.createdAt,
        "createdBy" to room.createdBy,
    ))

    fun findRoom(roomId: String, from: String?): Room {
        return SendBirdCall.getCachedRoomById(roomId) ?: throw RNCallsInternalError(from, RNCallsInternalError.Type.NOT_FOUND_ROOM)
    }

}
